llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeFlashAttentionPass.cpp

```diff
@@ -0,0 +1,420 @@
+    SmallVector<OpFoldResult> resultShape{sequenceTileLength,
+                                          sequenceTileLength};
+    Value emptySquare =
+        rewriter.create<tensor::EmptyOp>(loc, resultShape, elementType);
+    auto tensorType = RankedTensorType::get(
+        SmallVector<int64_t>(2, ShapedType::kDynamic), elementType);
+    Value qkTranspose =
+        computeQKTranspose(querySlice, keySlice, empty, emptySquare, zeroF32,
+                           tensorType, loc, rewriter);
+
+    empty = rewriter.create<tensor::EmptyOp>(
+        loc, SmallVector<OpFoldResult>{sequenceTileLength}, elementType);
+
+    // Compute current statistics
+    Value currentMax = computeRowwiseReduction<arith::MaxFOp>(
+        qkTranspose, empty, loc, rewriter);
+    Value currentSum = computeRowwiseReduction<arith::AddFOp>(
+        qkTranspose, empty, loc, rewriter);
+
+    // Update global statistics
+    Value newMax = computeNewMax(iterArgMax, currentMax, empty, loc, rewriter);
+    Value oldWeight =
+        subtractAndExponentiate1D(iterArgMax, newMax, empty, loc, rewriter);
+    Value currentWeight =
+        subtractAndExponentiate1D(currentMax, newMax, empty, loc, rewriter);
+    Value newSum = computeNewSum(iterArgSum, currentSum, oldWeight,
+                                 currentWeight, empty, loc, rewriter);
+
+    // Compute softmax
+    Value softmax = computeSoftmax(qkTranspose, currentMax, currentWeight,
+                                   newSum, emptySquare, loc, rewriter);
+
+    // Update accumulator
+    empty = rewriter.create<tensor::EmptyOp>(
+        loc, SmallVector<OpFoldResult>{sequenceLength, headDimension},
+        elementType);
+    Value scaledAcc = scaleAccumulator(outputSlice, iterArgSum, newSum,
+                                       oldWeight, empty, loc, rewriter);
+
+    // Compute matmul(softmax, v)
+    Value result = rewriter
+                       .create<linalg::MatmulOp>(
+                           loc, outputSlice.getType(),
+                           ValueRange{softmax, valueSlice}, scaledAcc)
+                       .getResult(0);
+
+    // Insert slices
+    auto [updatedAcc, updatedMax, updatedSum] =
+        insertSlices(result, iterArgResult, newMax, max, newSum, sum,
+                     queryShape, ivs, sequenceTileLength, loc, rewriter);
+
+    if (scf::YieldOp yieldOp = dyn_cast<scf::YieldOp>(
+            secondLoopNest.loops.back().getBody()->getTerminator())) {
+      rewriter.replaceOpWithNewOp<scf::YieldOp>(
+          yieldOp, ValueRange{updatedAcc, updatedMax, updatedSum});
+    }
+
+    if (scf::YieldOp yieldOp = dyn_cast<scf::YieldOp>(
+            firstLoopNest.loops.back().getBody()->getTerminator())) {
+      rewriter.setInsertionPoint(yieldOp);
+      rewriter.replaceOpWithNewOp<scf::YieldOp>(
+          yieldOp, ValueRange{secondLoopNest.results[0]});
+    }
+
+    attnOp.getResults()[0].replaceAllUsesWith(firstLoopNest.results[0]);
```

Comment:
Direct IR manipulations such as RAUW are disallowed in rewrite patterns. This may work accidentally, but may lead to catastrophic issues if this pattern is combined with other patterns or if the logic of the rewriter changes. This should use rewriter APIs instead.

Reply:
+1

This needs an explicit enumeration of uses and perform replacement under an updateRootInPlace or startRootUpdate/finalizeRootUpdate pair.

There is very unfortunately zero API support to guide users to do the right things atm.
We unforunately have such anti-patterns in SCF and other places .. 

see my comment here: https://discourse.llvm.org/t/rfc-canonicalizerpass-convergence-error-handling/67333/32

Reply:
I see thanks! I was not aware of that. Will modify as per Nicolas' suggestions.

Reply:
I was looking at the source for RAUW 
```
  /// Find uses of `from` and replace it with `to`
  void RewriterBase::replaceAllUsesWith(Value from, Value to) {
    for (OpOperand &operand : llvm::make_early_inc_range(from.getUses())) {
      Operation *op = operand.getOwner();
      updateRootInPlace(op, [&]() { operand.set(to); });
    }
  }
```
This looks exactly like Nicolas' suggestion? Or am I missing something? Why is RAUW unsafe while looping over uses + updateRootInPlace safe?

Reply:
Yes, this is the right thing to use. It was added recently and I don't know if IREE's version of LLVM has caught up with it.

This implementation of this new method is similar to RAUW but additionally calls `updateRootInPlace`. This notifies the derived rewriter that an op was modified. Different rewriters may be using that for different purposes, typically updating some internal data structures. Not notifying them about the change will make that internal state invalid wrt the IR. Specifically, the greedy rewriter (depending on the options) will consider the ops it knows were modified _again_, in case modifications allowed more patterns to apply. Without notification, it wouldn't and the result would be different. This happens to work stably because the flow here doesn't actually need the fixed-point greedy rewriter like pointed below. This isn't catastrophic, but other rewriters, such as the tracking one, may have a pointer to ops that satisfy certain invariants that will be violated by unreported in-place modifications, leading to a nasty crash.

Reply:
Thanks for the explanation. Also reminds me that I need to go back and re-read the source of greedy rewriter :)

Reply:
It's being actively worked on upstream, so maybe give it some time to stabilize.